### PR TITLE
Rest API block

### PR DESCRIPTION
**Pull request by markosly**
This block hosts a rest API server which allows getting and setting variables and calling funtions.

### CODE REVIEW

File: gr-azure-software-radio/python/rest_api.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                    return {"Error": f"Func: {name} not found"}
+                try:
+                    func = getattr(tbself, name)
+                    func(set_dict[name])
+                except Exception:
+                    return {'Error':f'Could not set {name}'}
+            return {"Ran":"ok"}
+
+        self.server_thread = threading.Thread(target=uvicorn.run, args=(app,), kwargs={'port':port}, daemon=True)
+        self.server_thread.start()
```
Comment (pomeroy3) [inline gr-azure-software-radio/python/rest_api.py:66]: Does this shut down cleanly when we stop the flowgraph, and can we use this to stop the flowgraph (and itself?)

It looks like other folks have taken stabs at shutting down cleanly here: https://github.com/encode/uvicorn/discussions/1103#discussioncomment-941726

File: gr-azure-software-radio/python/qa_rest_api.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        self.assertIsNotNone(instance)
+
+    def test_get_vars(self):
+        self.check_var1 = 1
+        self.check_var2 = '2'
+        self.check_var3 = {'pi':3.14}
+        port = self.get_free_port()
+        instance = RestApi(self, port)
+        # give server a second to startup
+        time.sleep(1)
```
Comment (pomeroy3) [inline gr-azure-software-radio/python/qa_rest_api.py:35]: What do you think about using the fastAPI TestClient for very focused tests on our endpoints? Might help with simplifying some of the low level tests, ensuring we return appropriate error codes, etc

File: gr-azure-software-radio/grc/azure_software_radio_rest_api.block.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+category: '[Azure software radio]'
+
+parameters:
+-   id: value
+    dtype: string
+    default: '0'
+    hide: all
+-   id: port
+    dtype: int
+    default: 8000
```
Comment (pomeroy3) [inline gr-azure-software-radio/grc/azure_software_radio_rest_api.block.yml:12]: Could we add in the ability to set a list of readable params, and a list of writable params, just to let users decide what parameters should be exposed, and in what way?

File: gr-azure-software-radio/python/qa_rest_api.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    def test_get_by_name(self):
+        self.pi_val = 3.14
+        port = self.get_free_port()
+        instance = RestApi(self, port)
+        # give server a second to startup
+        time.sleep(1)
+        para = "name=pi_val"
+        addr_str = 'http://127.0.0.1:' + str(port) + '/get-by-name'
+        resp = requests.get(addr_str, params=para)
+        res = eval(resp.text)
```
Comment (pomeroy3) [inline gr-azure-software-radio/python/qa_rest_api.py:52]: Could we be using resp.json() instead of eval(resp.text)?

**Review by pomeroy3 — CHANGES_REQUESTED** [review]
Had some questions on use of eval() in tests, suggestions on possible ways to make our tests run faster, and suggest that we add ways to limit what parameters are readable, and what are writeable.

We might want to look at how to let users limit what functions can be called as well

File: gr-azure-software-radio/python/rest_api.py (hunk 1 above)
Comment (markosly) [reply gr-azure-software-radio/python/rest_api.py:66]: Yes, in my testing using the daemon=True option gives the desired result here.

**Comment by markosly** [comment]
On the issue of the security holes presented by this implementation, I would like to propose creating 3 new parameters "callables", "readables" and "setables" the first being a list of the only functions which could be called by this api; any function not in the list would not be callable, and the same for readables being the only readable variables and setables being the only setable variables.

**Pull request closed without merging**